File: src/topological_sort/topological_sort.py
class Node:
    def __init__(self, value, adj):
        self.value = value
        self.adj = adj
        self.visited = False

def topological_sort(graph):
    sort_list = []
    for source in graph:
        source.visited = True
        topological_sort_node(source, sort_list)
    return sort_list

def topological_sort_node(node, sort_list):
    for e in node.adj:
        if not e.visited:
            e.visited = True
            topological_sort_node(e)
    sort_list.insert(0, node.value)

File: src/topological_sort/topological_sort.hs
import Data.List

data Node = Node { value :: Int, adj :: [Node] } deriving Eq

instance Show Node where
    show node = show $ value node

-- source node의 집합
type Graph = [Node]

topological_sort :: Graph -> [Node]
topological_sort graph = reverse . concat $ [topological_sort_node x [] | x <-graph]
    where topological_sort_node source visited = (concat [topological_sort_node a (source:visited) | a <- adj source, a `notElem` visited]) ++ [source]